# Understanding Logistic Regression Tables

Using the same code as in the previous exercise, try to interpret the summary table.

### More information about the dataset: 
Note that <i> interest rate</i> indicates the 3-month interest rate between banks and <i> duration </i> indicates the time since the last contact was made with a given consumer. The <i> previous </i> variable shows whether the last marketing campaign was successful with this customer. The <i>March</i> and <i> May </i> are Boolean variables that account for when the call was made to the specific customer and <i> credit </i> shows if the customer has enough credit to avoid defaulting.

<i> Notes: 
    <li> the first column of the dataset is an index one; </li>
    <li> you don't need the graph for this exercise; </li>
    <li> the dataset used is much bigger </li>
</i>

## Import the relevant libraries

In [1]:
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from scipy import stats
stats.chisqprob = lambda chisq, df: stats.chi2.sf(chisq, df)

## Load the data

Load the ‘Bank_data.csv’ dataset.

In [3]:
raw_data = pd.read_csv('Bank-data.csv')
raw_data

Unnamed: 0  interest_rate  credit  march  may  previous  duration    y
0             0          1.334     0.0    1.0  0.0       0.0     117.0   no
1             1          0.767     0.0    0.0  2.0       1.0     274.0  yes
2             2          4.858     0.0    1.0  0.0       0.0     167.0   no
3             3          4.120     0.0    0.0  0.0       0.0     686.0  yes
4             4          4.856     0.0    1.0  0.0       0.0     157.0   no
5             5          0.899     0.0    0.0  1.0       0.0     126.0   no
6             6          4.962     0.0    0.0  0.0       0.0      84.0   no
7             7          4.858     0.0    1.0  0.0       0.0      17.0   no
8             8          4.962     0.0    0.0  0.0       0.0     704.0  yes
9             9          4.865     0.0    0.0  0.0       0.0     185.0   no
10           10          1.365     0.0    0.0  1.0       1.0     374.0   no
11           11          0.773     0.0    0.0  0.0       0.0      91.0  yes
12           12          0.714     0.0    0.0  2.0       1.0     169.0  yes
13           13          4.864     0.0    0.0  0.0       0.0     249.0   no
14           14          4.966     0.0    0.0  0.0       0.0     215.0   no
15           15          0.904     0.0    0.0  0.0       0.0     324.0  yes
16           16          0.849     0.0    0.0  2.0       1.0     159.0  yes
17           17          1.811     1.0    0.0  0.0       0.0     120.0  yes
18           18          1.264     0.0    1.0  0.0       0.0     337.0  yes
19           19          4.076     0.0    0.0  0.0       0.0     640.0   no
20           20          1.262     0.0    0.0  0.0       0.0     663.0  yes
21           21          0.695     0.0    1.0  2.0       1.0     403.0  yes
22           22          4.960     0.0    0.0  0.0       0.0     300.0   no
23           23          4.963     0.0    0.0  0.0       0.0     255.0  yes
24           24          1.354     0.0    1.0  1.0       1.0     293.0  yes
25           25          1.244     0.0    1.0  0.0       0.0      68.0   no
26           26          0.748     0.0    0.0  1.0       0.0     266.0  yes
27           27          0.878     0.0    0.0  0.0       0.0     272.0  yes
28           28          0.644     0.0    0.0  0.0       0.0     398.0  yes
29           29          4.968     0.0    0.0  0.0       0.0     126.0   no
..          ...            ...     ...    ...  ...       ...       ...  ...
488         488          0.885     0.0    0.0  2.0       1.0     608.0  yes
489         489          4.153     0.0    0.0  0.0       0.0     250.0   no
490         490          0.735     0.0    0.0  0.0       0.0     107.0  yes
491         491          4.965     0.0    0.0  0.0       0.0     716.0  yes
492         492          4.859     0.0    1.0  0.0       0.0     619.0  yes
493         493          4.864     0.0    0.0  0.0       0.0     406.0   no
494         494          4.963     0.0    0.0  0.0       0.0      73.0   no
495         495          1.334     0.0    1.0  0.0       0.0     127.0   no
496         496          4.153     0.0    0.0  0.0       0.0     187.0   no
497         497          0.646     1.0    0.0  1.0       1.0     180.0  yes
498         498          1.266     0.0    1.0  0.0       0.0     326.0  yes
499         499          1.405     0.0    0.0  1.0       0.0     109.0  yes
500         500          0.900     0.0    0.0  2.0       1.0     470.0  yes
501         501          4.076     0.0    0.0  0.0       0.0      73.0   no
502         502          1.050     0.0    0.0  0.0       0.0     141.0   no
503         503          1.029     0.0    0.0  1.0       0.0     115.0   no
504         504          0.748     0.0    0.0  1.0       1.0     171.0   no
505         505          4.961     0.0    0.0  0.0       0.0     498.0   no
506         506          1.268     0.0    0.0  0.0       0.0     365.0  yes
507         507          4.959     0.0    0.0  0.0       0.0      10.0   no
508         508          4.021     0.0    0.0  1.0   

In [4]:
# We make sure to create a copy of the data before we start altering it. Note that we don't change the original data we loaded.
data = raw_data.copy()

# Removes the index column thata comes with the data
data = data.drop(['Unnamed: 0'], axis = 1)

# We use the map function to change any 'yes' values to 1 and 'no'values to 0. 
data['y'] = data['y'].map({'yes':1, 'no':0})
data

interest_rate  credit  march  may  previous  duration  y
0            1.334     0.0    1.0  0.0       0.0     117.0  0
1            0.767     0.0    0.0  2.0       1.0     274.0  1
2            4.858     0.0    1.0  0.0       0.0     167.0  0
3            4.120     0.0    0.0  0.0       0.0     686.0  1
4            4.856     0.0    1.0  0.0       0.0     157.0  0
5            0.899     0.0    0.0  1.0       0.0     126.0  0
6            4.962     0.0    0.0  0.0       0.0      84.0  0
7            4.858     0.0    1.0  0.0       0.0      17.0  0
8            4.962     0.0    0.0  0.0       0.0     704.0  1
9            4.865     0.0    0.0  0.0       0.0     185.0  0
10           1.365     0.0    0.0  1.0       1.0     374.0  0
11           0.773     0.0    0.0  0.0       0.0      91.0  1
12           0.714     0.0    0.0  2.0       1.0     169.0  1
13           4.864     0.0    0.0  0.0       0.0     249.0  0
14           4.966     0.0    0.0  0.0       0.0     215.0  0
15           0.904     0.0    0.0  0.0       0.0     324.0  1
16           0.849     0.0    0.0  2.0       1.0     159.0  1
17           1.811     1.0    0.0  0.0       0.0     120.0  1
18           1.264     0.0    1.0  0.0       0.0     337.0  1
19           4.076     0.0    0.0  0.0       0.0     640.0  0
20           1.262     0.0    0.0  0.0       0.0     663.0  1
21           0.695     0.0    1.0  2.0       1.0     403.0  1
22           4.960     0.0    0.0  0.0       0.0     300.0  0
23           4.963     0.0    0.0  0.0       0.0     255.0  1
24           1.354     0.0    1.0  1.0       1.0     293.0  1
25           1.244     0.0    1.0  0.0       0.0      68.0  0
26           0.748     0.0    0.0  1.0       0.0     266.0  1
27           0.878     0.0    0.0  0.0       0.0     272.0  1
28           0.644     0.0    0.0  0.0       0.0     398.0  1
29           4.968     0.0    0.0  0.0       0.0     126.0  0
..             ...     ...    ...  ...       ...       ... ..
488          0.885     0.0    0.0  2.0       1.0     608.0  1
489          4.153     0.0    0.0  0.0       0.0     250.0  0
490          0.735     0.0    0.0  0.0       0.0     107.0  1
491          4.965     0.0    0.0  0.0       0.0     716.0  1
492          4.859     0.0    1.0  0.0       0.0     619.0  1
493          4.864     0.0    0.0  0.0       0.0     406.0  0
494          4.963     0.0    0.0  0.0       0.0      73.0  0
495          1.334     0.0    1.0  0.0       0.0     127.0  0
496          4.153     0.0    0.0  0.0       0.0     187.0  0
497          0.646     1.0    0.0  1.0       1.0     180.0  1
498          1.266     0.0    1.0  0.0       0.0     326.0  1
499          1.405     0.0    0.0  1.0       0.0     109.0  1
500          0.900     0.0    0.0  2.0       1.0     470.0  1
501          4.076     0.0    0.0  0.0       0.0      73.0  0
502          1.050     0.0    0.0  0.0       0.0     141.0  0
503          1.029     0.0    0.0  1.0       0.0     115.0  0
504          0.748     0.0    0.0  1.0       1.0     171.0  0
505          4.961     0.0    0.0  0.0       0.0     498.0  0
506          1.268     0.0    0.0  0.0       0.0     365.0  1
507          4.959     0.0    0.0  0.0       0.0      10.0  0
508          4.021     0.0    0.0  1.0       0.0     796.0  1
509          0.877     0.0    0.0  2.0       1.0     279.0  1
510          1.327     0.0    1.0  0.0       0.0     476.0  1
511          4.965     0.0    0.0  0.0       0.0     479.0  1
512          1.266     0.0    1.0  1.0       0.0     225.0  0
513          1.334     0.0    1.0  0.0       0.0     204.0  0
514          0.861     0.0    0.0  2.0       1.0     806.0  1
515          0.879     0.0    0.0  0.0       0.0     290.0  0
516          0.877     0.0    0.0  5.0       1.0     473.0  1
517          4.965     0.0    0.0  0.0       0.0     142.0  0

[518 rows x 7 columns]

In [5]:
data.describe()

interest_rate      credit       march         may    previous  \
count     518.000000  518.000000  518.000000  518.000000  518.000000   
mean        2.835776    0.034749    0.266409    0.388031    0.127413   
std         1.876903    0.183321    0.442508    0.814527    0.333758   
min         0.635000    0.000000    0.000000    0.000000    0.000000   
25%         1.042750    0.000000    0.000000    0.000000    0.000000   
50%         1.466000    0.000000    0.000000    0.000000    0.000000   
75%         4.956500    0.000000    1.000000    0.000000    0.000000   
max         4.970000    1.000000    1.000000    5.000000    1.000000   

          duration           y  
count   518.000000  518.000000  
mean    382.177606    0.500000  
std     344.295990    0.500483  
min       9.000000    0.000000  
25%     155.000000    0.000000  
50%     266.500000    0.500000  
75%     482.750000    1.000000  
max    2653.000000    1.000000

### Declare the dependent and independent variables

Use 'duration' as the independet variable.

In [6]:
y = data['y']
x1 = data['duration']

### Simple Logistic Regression

Run the regression.

In [7]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()

Optimization terminated successfully.
         Current function value: 0.546118
         Iterations 7


### Interpretation

In [8]:
results_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                  518
Model:                          Logit   Df Residuals:                      516
Method:                           MLE   Df Model:                            1
Date:                Sat, 15 Dec 2018   Pseudo R-squ.:                  0.2121
Time:                        23:27:21   Log-Likelihood:                -282.89
converged:                       True   LL-Null:                       -359.05
                                        LLR p-value:                 5.387e-35
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.7001      0.192     -8.863      0.000      -2.076      -1.324
duration       0.0051      0.001      9.159      0.000       0.004       0.006
==============================================================================
"""

The dependent variable is 'duration'. The model used is a Logit regression (logistic in common lingo), while the method - Maximum Likelihood Estimation (MLE). It has clearly converged after classifyin 518 observations. 

The Pseudo R-squared is 0.21 which is within the 'acceptable region'. 

The duration variable is significant and its coefficient is 0.0051.

The constant is also significant and equals: -1.70